# Serie 13
## Aufgabe 13.1
In der Bibliothek `ISLR` hat es den Datensatz **Carseats**. Wir möchten `Sales` (Anzahl Kinderautositze) aufgrund von verschiedenen Prädiktoren in 400 verschiedenen Standorten vorhersagen.  
Der Datensatz enthält qualitative Prädiktoren, wie `ShelveLoc` als Indikator der Lage im Gestell, das heisst der Platz in einem Geschäft, wo der Autositz ausgestellt ist. Der Prädiktor nimmt die drei Werte `Bad`, `Medium` und `Good` an. Für qualitative Variablen generiert `R` Dummy-Variablen automatisch.

a) Untersuchen Sie den Datensatz mit `head(Carseat)` und `?Carseat`.

In [1]:
library(ISLR)
head(Carseats)
help(Carseats)

Sales CompPrice Income Advertising Population Price ShelveLoc Age Education
1  9.50 138        73    11          276        120   Bad       42  17       
2 11.22 111        48    16          260         83   Good      65  10       
3 10.06 113        35    10          269         80   Medium    59  12       
4  7.40 117       100     4          466         97   Medium    55  14       
5  4.15 141        64     3          340        128   Bad       38  13       
6 10.81 124       113    13          501         72   Bad       78  16       
  Urban US 
1 Yes   Yes
2 Yes   Yes
3 Yes   Yes
4 Yes   Yes
5 Yes   No 
6 No    Yes

Carseats                 package:ISLR                  R Documentation

_S_a_l_e_s _o_f _C_h_i_l_d _C_a_r _S_e_a_t_s

_D_e_s_c_r_i_p_t_i_o_n:

     A simulated data set containing sales of child car seats at 400
     different stores.

_U_s_a_g_e:

     Carseats
     
_F_o_r_m_a_t:

     A data frame with 400 observations on the following 11 variables.

     ‘Sales’ Unit sales (in thousands) at each location

     ‘CompPrice’ Price charged by competitor at each location

     ‘Income’ Community income level (in thousands of dollars)

     ‘Advertising’ Local advertising budget for company at each
          location (in thousands of dollars)

     ‘Population’ Population size in region (in thousands)

     ‘Price’ Price company charges for car seats at each site

     ‘ShelveLoc’ A factor with levels ‘Bad’, ‘Good’ and ‘Medium’
          indicating the quality of the shelving location for the car
          seats at each site

     ‘Age’ Average age of the local population

     ‘Education’ Education level at each location

     ‘Urban’ A factor with levels ‘No’ and ‘Yes’ to indicate whether
          the store is in an urban or rural location

     ‘US’ A factor with levels ‘No’ and ‘Yes’ to indicate whether the
          store is in the US or not

_S_o_u_r_c_e:

     Simulated data

_R_e_f_e_r_e_n_c_e_s:

     James, G., Witten, D., Hastie, T., and Tibshirani, R. (2013) _An
     Introduction to Statistical Learning with applications in R_,
     <URL: www.StatLearning.com>, Springer-Verlag, New York

_E_x_a_m_p_l_e_s:

     summary(Carseats)
     lm.fit=lm(Sales~Advertising+Price,data=Carseats)

b) Finden Sie mit `lm()` ein multiples Regressionsmodell um `Sales` aus `Price`, `Urban` und `US` vorherzusagen.

In [2]:
fit <- lm(Sales ~ Price + Urban + US, data = Carseats)
summary(fit)


Call:
lm(formula = Sales ~ Price + Urban + US, data = Carseats)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.9206 -1.6220 -0.0564  1.5786  7.0581 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 13.043469   0.651012  20.036  < 2e-16 ***
Price       -0.054459   0.005242 -10.389  < 2e-16 ***
UrbanYes    -0.021916   0.271650  -0.081    0.936    
USYes        1.200573   0.259042   4.635 4.86e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.472 on 396 degrees of freedom
Multiple R-squared:  0.2393,	Adjusted R-squared:  0.2335 
F-statistic: 41.52 on 3 and 396 DF,  p-value: < 2.2e-16


c) Interpretieren Sie die Koeffizienten in diesem Modell. Achten Sie darauf, dass einige Variablen qualitativ sind.

* 13.04 ist schwierig zu interpretieren. Mittelwert aller Verkäufe, auch bei einem Preis von \$0.
* nimmt der Preis um 1 Dollar zu, nehmen die Verkäufe um $-0.05$ Einheiten ab.
* Der Verkauft in ländlichen Gebieten ist eher kleiner. $p$-Wert aber ist aber nicht signifikant $\Rightarrow$ zufällig
* in den USA werden mehr Sitze gekauft, als anderswo

`Price`und `US` sind signifikant, ob Autositz in `Urban` gekauft wird hat keinen Zusammenhang mit `Sales`

d) Schreiben Sie das Modell in Gleichungsform. Achten Sie darauf, dass Sie die qualitativen Variablen richtig behandeln.

$$Urban = x_{2i}= \cases{
      0, & falls $i$-te Person in der Stadt lebt \\
      1, & falls $i$-te Person auf dem Land lebt}$$
$$US = x_{3i}= \cases{
0, & falls $i$-te Person in der USA lebt \\
1, & falls $i$-te Person nicht in der USA lebt}$$
$$Sales =\beta_0 + \beta_1 * Price + \beta_2 * x_{2i} + \beta_3 * x_{3i}$$

e) Für welche Prädiktoren kann die Nullhypothese $H_0 : \beta_j = 0$ verworfen werden?

alle ausser `Urban`, die ist statistisch nicht signifikant.

f) Auf der Basis der vorhergehenden Frage, finden Sie ein kleineres Modell, das nur Prädiktoren verwendet für die es Hinweise auf einen Zusammenhang mit der Zielvariablen gibt.

In [3]:
fitterFit <- lm(Sales ~ Price + US, data = Carseats)
summary(fitterFit)


Call:
lm(formula = Sales ~ Price + US, data = Carseats)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.9269 -1.6286 -0.0574  1.5766  7.0515 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 13.03079    0.63098  20.652  < 2e-16 ***
Price       -0.05448    0.00523 -10.416  < 2e-16 ***
USYes        1.19964    0.25846   4.641 4.71e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.469 on 397 degrees of freedom
Multiple R-squared:  0.2393,	Adjusted R-squared:  0.2354 
F-statistic: 62.43 on 2 and 397 DF,  p-value: < 2.2e-16


g) Wie genau passen die Modelle in a) und e) die Daten an?  

Die Variablen haben einen statistisch signifikanten Zusammenhang. Allerdings ist der $R^2$-Wert schlecht. Die Daten erklären das Modell schlecht.

## Aufgabe 13.2
Wir führen noch eine multiple lineare Regression für Auto aus der letzten Übung durch.

a) Bestimmen Sie mit `regsubset` mit Option `forward`, das multiple Regressionsmodell mit zwei Prädiktoren und bestimmen Sie die Koeffizienten diese Modelles. Interpretieren Sie die Koeffizienten.

In [2]:
library(ISLR)
library(leaps)
reg <- regsubsets(mpg ~ ., data = Auto[ , -9], method="forward")
reg.sum <- summary(reg)
reg.sum$which

(Intercept) cylinders displacement horsepower weight acceleration year 
1 TRUE        FALSE     FALSE        FALSE      TRUE   FALSE        FALSE
2 TRUE        FALSE     FALSE        FALSE      TRUE   FALSE         TRUE
3 TRUE        FALSE     FALSE        FALSE      TRUE   FALSE         TRUE
4 TRUE        FALSE      TRUE        FALSE      TRUE   FALSE         TRUE
5 TRUE        FALSE      TRUE         TRUE      TRUE   FALSE         TRUE
6 TRUE         TRUE      TRUE         TRUE      TRUE   FALSE         TRUE
7 TRUE         TRUE      TRUE         TRUE      TRUE    TRUE         TRUE
  origin
1 FALSE 
2 FALSE 
3  TRUE 
4  TRUE 
5  TRUE 
6  TRUE 
7  TRUE

In [5]:
fitCar <- lm(mpg ~ weight + year, data=Auto)
summary(fitCar)$coef

Estimate      Std. Error   t value    Pr(>|t|)     
(Intercept) -14.347253018 4.0065185631  -3.580978  3.856624e-04
weight       -0.006632075 0.0002145559 -30.910708 8.361624e-107
year          0.757318281 0.0494726873  15.307806  9.772260e-42

In [6]:
fitCar <- lm(mpg ~ weight * year, data=Auto)
summary(fitCar)$coef

Estimate      Std. Error   t value   Pr(>|t|)    
(Intercept) -1.104519e+02 1.294697e+01 -8.531103 3.295213e-16
weight       2.754652e-02 4.413436e-03  6.241514 1.136624e-09
year         2.040476e+00 1.718209e-01 11.875599 5.881890e-28
weight:year -4.579323e-04 5.907150e-05 -7.752171 8.015486e-14

* Die Reichweite nimmt mit dem steigendem Gewicht ab
* Je älter das Auto umso weniger Reichweite pro Gallone

b) Hätte es einen Unterschied gegeben, wenn wir die Option backward gewählt hätten?

In [7]:
regB <- regsubsets(mpg ~ ., data = Auto[ , -9], method="backward")
regB.sum <- summary(regB)
regB.sum$which

(Intercept) cylinders displacement horsepower weight acceleration year 
1 TRUE        FALSE     FALSE        FALSE      TRUE   FALSE        FALSE
2 TRUE        FALSE     FALSE        FALSE      TRUE   FALSE         TRUE
3 TRUE        FALSE     FALSE        FALSE      TRUE   FALSE         TRUE
4 TRUE        FALSE      TRUE        FALSE      TRUE   FALSE         TRUE
5 TRUE        FALSE      TRUE         TRUE      TRUE   FALSE         TRUE
6 TRUE         TRUE      TRUE         TRUE      TRUE   FALSE         TRUE
7 TRUE         TRUE      TRUE         TRUE      TRUE    TRUE         TRUE
  origin
1 FALSE 
2 FALSE 
3  TRUE 
4  TRUE 
5  TRUE 
6  TRUE 
7  TRUE

## Test zu Rückwärtsselektion

In [15]:
f.full <- lm(mpg ~ ., data = Auto[ , -9])
f.empty <- lm(mpg ~ NULL, data = Auto[ , -9])
summary(f.full)


Call:
lm(formula = mpg ~ ., data = Auto[, -9])

Residuals:
    Min      1Q  Median      3Q     Max 
-9.5903 -2.1565 -0.1169  1.8690 13.0604 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -17.218435   4.644294  -3.707  0.00024 ***
cylinders     -0.493376   0.323282  -1.526  0.12780    
displacement   0.019896   0.007515   2.647  0.00844 ** 
horsepower    -0.016951   0.013787  -1.230  0.21963    
weight        -0.006474   0.000652  -9.929  < 2e-16 ***
acceleration   0.080576   0.098845   0.815  0.41548    
year           0.750773   0.050973  14.729  < 2e-16 ***
origin         1.426141   0.278136   5.127 4.67e-07 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.328 on 384 degrees of freedom
Multiple R-squared:  0.8215,	Adjusted R-squared:  0.8182 
F-statistic: 252.4 on 7 and 384 DF,  p-value: < 2.2e-16


In [16]:
drop1(f.full, scope = f.full, trace = TRUE)

Df Sum of Sq   RSS      AIC      
<none>       NA          NA 4252.213  950.5017
cylinders     1   25.791491 4278.004  950.8722
displacement  1   77.612668 4329.825  955.5921
horsepower    1   16.739754 4268.952  950.0419
weight        1 1091.631693 5343.844 1038.0760
acceleration  1    7.358417 4259.571  949.1795
year          1 2402.249906 6654.462 1124.0582
origin        1  291.134494 4543.347  974.4617

In [17]:
f.7 <- update(f.full, . ~. - acceleration)
drop1(f.7, scope = f.7, trace = TRUE)

Df Sum of Sq  RSS      AIC      
<none>       NA         NA 4259.571  949.1795
cylinders     1   27.27126 4286.842  949.6812
displacement  1   73.57187 4333.143  953.8923
horsepower    1   53.80189 4313.373  952.0997
weight        1 1310.43126 5570.002 1052.3244
year          1 2396.17365 6655.745 1122.1337
origin        1  292.01817 4551.589  973.1722